rpmio/rpmkeyring.c

```diff
@@ -153,37 +155,72 @@ rpmPubkey rpmPubkeyNew(const uint8_t *pkt, size_t pktlen)
     return key;
 }
 
-static rpmPubkey rpmPubkeyNewSubkey(pgpDigParams pgpkey)
+static rpmPubkey rpmPubkeyNewSubkey(rpmPubkey primarykey, pgpDigParams pgpkey)
 {
     rpmPubkey key = new rpmPubkey_s {};
-    /* Packets with all subkeys already stored in main key */
+    /* Packets with all subkeys already stored in primary key */
+    key->primarykey = primarykey;
     key->pgpkey = pgpkey;
     key->keyid = key2str(pgpDigParamsSignID(pgpkey));
     key->nrefs = 1;
     return key;
 }
 
-rpmPubkey *rpmGetSubkeys(rpmPubkey mainkey, int *count)
+rpmPubkey *rpmGetSubkeys(rpmPubkey primarykey, int *count)
 {
     rpmPubkey *subkeys = NULL;
     pgpDigParams *pgpsubkeys = NULL;
     int pgpsubkeysCount = 0;
     int i;
 
-    if (mainkey && !pgpPrtParamsSubkeys(mainkey->pkt.data(), mainkey->pkt.size(),
-			mainkey->pgpkey, &pgpsubkeys, &pgpsubkeysCount)) {
+    wrlock lock(primarykey->mutex);
+
+    if (primarykey && !pgpPrtParamsSubkeys(primarykey->pkt.data(), primarykey->pkt.size(),
+			primarykey->pgpkey, &pgpsubkeys, &pgpsubkeysCount)) {
 
 	/* Returned to C, can't use new */
 	subkeys = (rpmPubkey *)xmalloc(pgpsubkeysCount * sizeof(*subkeys));
-	for (i = 0; i < pgpsubkeysCount; i++)
-	    subkeys[i] = rpmPubkeyNewSubkey(pgpsubkeys[i]);
+	for (i = 0; i < pgpsubkeysCount; i++) {
+	    subkeys[i] = rpmPubkeyNewSubkey(primarykey, pgpsubkeys[i]);
+	    primarykey->nrefs++; /* can't use rpmPubkeyLink due to wrlock */
```

Comment:
Like I said in ... chat I think, in a case like this you'll just make a private unlocked version of the function you need. See macros.c for an example: there's an "External interfaces" section towards the bottom that lock on entry, and everything else operates assumes that locks are handled.

Reply:
Yes you want because then you don't need the comment to explain it, and it's the right technique for this kind of  job. Wrapping it into function was is far less text than arguing it :sweat_smile:

Reply:
Done. Not quite sure if the name is that great. Do we have a naming scheme for that? Just `_intern`? or `_internal`? or should I just keep `Prelocked`?

Reply:
The common practise across the codebase is to leave the rpm-prefix out of such internal functions (rpmstrpool has examples too), so it'd be just pubkeyLink(). And those commonly don't check for NULL either, it's expected that the caller knows what they're doing.

Reply:
Doh... ofc I have seen that over and over again. But yesterday my mind was looking for a name to append and couldn't remember one - for good reason. Changed the name.